pygsp/graphs/graph.py

```diff
@@ -122,6 +126,238 @@ def __repr__(self, limit=None):
+        fmt : String
+            Format in which the graph is encoded. Currently supported format are:
+            GML and gpickle.
+        lib : String
+            Python library used in background to load the graph.
+            Supported library are networkx and graph_tool
+        
+        Returns
+        -------
+            g : :class:`~pygsp.graphs.Graph`
+
+        """
+        if fmt == 'auto':
+            fmt = path.split('.')[-1]
+
+        err = NotImplementedError('{} can not be load with {}. \
+               Try another background library'.format(fmt, lib))
+
+        if lib == 'networkx':
+            import networkx
+            if fmt == 'gml':
+                g = networkx.read_gml(path)
+                return cls.from_networkx(g)
+            if fmt in ['gpickle', 'p', 'pkl', 'pickle']:
+                g = networkx.read_gpickle(path)
+                return cls.from_networkx(g)
+            raise err
+        if lib == 'graph_tool':
+            import graph_tool
+            g = graph_tool.load_graph(path, fmt=fmt)
+            return cls.from_graphtool(g)
+
+        raise NotImplementedError('the format {} is not suported'.format(fmt))
+
+    def save(self, path, fmt='auto', lib='networkx'):
+        r"""Save the graph into a file
+        
+        Parameters
+        ----------
+        path : String
+            Where to save file on the disk.
+        fmt : String
+            Format in which the graph will be encoded. The format is guessed from
+            the `path` extention when fmt is set to 'auto'
+            Currently supported format are:
+            GML and gpickle.
+        lib : String
+            Python library used in background to save the graph.
+            Supported library are networkx and graph_tool
+        """
+        if fmt == 'auto':
+           fmt = path.split('.')[-1]
+
+        err = NotImplementedError('{} can not be save with {}. \
+                Try another background library'.format(fmt, lib))
+
+        if lib == 'networkx':
+            import networkx
+            if fmt == 'gml':
+                g = self.to_networkx()
+                networkx.write_gml(g, path)
+                return
+            raise err
+
+        if lib == 'graph_tool':
+            import graph_tool
+            g = self.to_graphtool()
+            g.save(path, fmt=fmt)
+            return
+
+        raise NotImplementedError('the format {} is not suported'.format(fmt))
+
+    def set_signal(self, signal, signal_name):
+        r"""
+        Add or modify a signal to the graph
+
+        Parameters
+        ----------
+        signal : numpy.array
+            An array maping from node to his value. For example the value of the singal at node i is signal[i]
+        signal_name : String
+            Name associated to the signal.
+        """
+        assert len(signal) == self.N, "A value must be attached to every vertex in the graph"
+        self.signals[signal_name] = np.array(signal)
```

Comment:
Should I keep it as `np.array` and put the `copy=False` as it seems to be the same regarding this [post](https://stackoverflow.com/questions/14415741/numpy-array-vs-asarray). I feel like this easier to understand the code that way.

Apart from here, I am using `np.array` in conversion methods like `to_graphtool()` and `from_networkx`. Should I make the change there as well or not. I think copying the array in those situations is good as you might want your `pyGSP`'s graph to be independent of your `networkx` or `graphtool` one. Changing one should not affect the other I think.

Reply:
I think `asarray` is pretty standard, but as you want.

I think we should copy as little as possible in general. Users can always copy with standard tools like `copy.deepcopy`, but it's hard to prevent copies to happen.